Support multi volume rar files

ShayArtzi opened the pull request:
Hi @ahadas!

I'm creating this PR to discuss some of the challenges in implementing multi volume support. I'm starting with RAR file as the 7z implementation will be different.

Some points I'd like to get your thought on if possible:

1. I created `SevenZipRarMultiVolumeCallbackHandler` to handle multi volume archives and used a regular expression on the file name to detect when it's right to use it. This can be optimized by not compiling the pattern each time but keeping it as a field in the class. What do you think about that approach?
2. I wanted to make sure all volume streams are closed at the end of reading operation, so I created a wrapper class `InArchiveWrapper` that carries over the original `IInArchive` and the handler that references other volume streams.
3. I needed a way to create an `AbstractFile` without recursively calling `SevenZipRarMultiVolumeCallbackHandler::getStream`. I couldn't find a proper way to do it so I changed the modifier of `FileFactory::createRawFile` to `public`. I guess that's not ideal so I'm open to other suggestions (I wanted to reuse the existing `SignatureCheckedRandomAccessFile` class)

Here is a [reference](https://sevenzipjbind.sourceforge.net/extraction_snippets.html#open-multipart-rar-archives) from `sevenzipjbind`'s docs about opening multi volume archives.

Using the code in this PR, I was able to extract a multi-volume RAR archive successfully.

Thanks very much :)

ahadas commented on sevenzipjbindings/src/main/java/com/mucommander/sevenzipjbindings/SevenZipJBindingROArchiveFile.java:
```diff
@@ -79,8 +77,20 @@ public SevenZipJBindingROArchiveFile check() throws IOException {
-            SignatureCheckedRandomAccessFile in = new SignatureCheckedRandomAccessFile(file, formatSignature);
-            inArchive = SevenZip.openInArchive(sevenZipJBindingFormat, in, password);
+            Pattern multiPartRarPattern = Pattern.compile("[.]part\\d+[.]rar");
+            Matcher matcher = multiPartRarPattern.matcher(file.getAbsolutePath());
```
we can match against the filename (rather than the whole path), right?

ShayArtzi replied:
sure

ahadas commented on sevenzipjbindings/src/main/java/com/mucommander/sevenzipjbindings/SevenZipJBindingROArchiveFile.java:
```diff
@@ -131,6 +142,11 @@ public InputStream getEntryInputStream(ArchiveEntry entry, ArchiveEntryIterator
+                    try {
+                        inArchive.close();
+                    } catch (SevenZipException e) {
+                        e.printStackTrace();
```
should use the logger instead

ShayArtzi replied:
done

ahadas commented on sevenzipjbindings/src/main/java/com/mucommander/sevenzipjbindings/multivolume/SevenZipRarMultiVolumeCallbackHandler.java:
```diff
@@ -0,0 +1,89 @@
+                stream.seek(0, ISeekableStream.SEEK_SET);
+            } else {
+                File tempFile = new File(filename);
+                AbstractFile abstractFile = FileFactory.createRawFile(FileURL.getFileURL(tempFile.getAbsolutePath()), null, new HashMap<>());
```
why not to use FileFactory#getFile?

ShayArtzi replied:
That's the thing - using this method takes me into an endless recursion. Look at the stack trace:
```2023-04-21 20:17:38.566 [FelixStartLevel] ERROR com.mucommander.Application:439 - Startup failed
java.lang.StackOverflowError: null
	at java.base/java.util.regex.Pattern$BmpCharProperty.match(Pattern.java:3966)
	at java.base/java.util.regex.Pattern$Curly.match(Pattern.java:4375)
	at java.base/java.util.regex.Pattern$Curly.match0(Pattern.java:4395)
	at java.base/java.util.regex.Pattern$Curly.match(Pattern.java:4382)
	at java.base/java.util.regex.Pattern$Begin.match(Pattern.java:3686)
	at java.base/java.util.regex.Matcher.match(Matcher.java:1756)
	at java.base/java.util.regex.Matcher.matches(Matcher.java:713)
	at com.mucommander.commons.file.protocol.local.LocalFile.isRoot(LocalFile.java:874)
	at com.mucommander.commons.file.protocol.local.LocalFile.getName(LocalFile.java:762)
	at com.mucommander.sevenzipjbindings.SevenZipJBindingROArchiveFile.openInArchive(SevenZipJBindingROArchiveFile.java:82)
	at com.mucommander.sevenzipjbindings.SevenZipJBindingROArchiveFile.check(SevenZipJBindingROArchiveFile.java:76)
	at com.mucommander.commons.file.archive.rar.RarFormatProvider.getFile(RarFormatProvider.java:56)
	at com.mucommander.commons.file.FileFactory.wrapArchive(FileFactory.java:682)
	at com.mucommander.commons.file.FileFactory.getFile(FileFactory.java:491)
	at com.mucommander.commons.file.FileFactory.getFile(FileFactory.java:402)
	at com.mucommander.commons.file.FileFactory.getFile(FileFactory.java:356)
	at com.mucommander.commons.file.FileFactory.getFile(FileFactory.java:314)
	at com.mucommander.sevenzipjbindings.multivolume.SevenZipRarMultiVolumeCallbackHandler.getStream(SevenZipRarMultiVolumeCallbackHandler.java:58)
	at com.mucommander.sevenzipjbindings.SevenZipJBindingROArchiveFile.openInArchive(SevenZipJBindingROArchiveFile.java:86)
	at com.mucommander.sevenzipjbindings.SevenZipJBindingROArchiveFile.check(SevenZipJBindingROArchiveFile.java:76)
	at com.mucommander.commons.file.archive.rar.RarFormatProvider.getFile(RarFormatProvider.java:56)
	at com.mucommander.commons.file.FileFactory.wrapArchive(FileFactory.java:682)
	at com.mucommander.commons.file.FileFactory.getFile(FileFactory.java:491)
	at com.mucommander.commons.file.FileFactory.getFile(FileFactory.java:402)
	at com.mucommander.commons.file.FileFactory.getFile(FileFactory.java:356)
	at com.mucommander.commons.file.FileFactory.getFile(FileFactory.java:314)
	at com.mucommander.sevenzipjbindings.multivolume.SevenZipRarMultiVolumeCallbackHandler.getStream(SevenZipRarMultiVolumeCallbackHandler.java:58)
	at com.mucommander.sevenzipjbindings.SevenZipJBindingROArchiveFile.openInArchive(SevenZipJBindingROArchiveFile.java:86)
	at com.mucommander.sevenzipjbindings.SevenZipJBindingROArchiveFile.check(SevenZipJBindingROArchiveFile.java:76)
	at com.mucommander.commons.file.archive.rar.RarFormatProvider.getFile(RarFormatProvider.java:56)
	at com.mucommander.commons.file.FileFactory.wrapArchive(FileFactory.java:682)
	at com.mucommander.commons.file.FileFactory.getFile(FileFactory.java:491)
	at com.mucommander.commons.file.FileFactory.getFile(FileFactory.java:402)
	at com.mucommander.commons.file.FileFactory.getFile(FileFactory.java:356)
	at com.mucommander.commons.file.FileFactory.getFile(FileFactory.java:314)
	at com.mucommander.sevenzipjbindings.multivolume.SevenZipRarMultiVolumeCallbackHandler.getStream(SevenZipRarMultiVolumeCallbackHandler.java:58)
	at com.mucommander.sevenzipjbindings.SevenZipJBindingROArchiveFile.openInArchive(SevenZipJBindingROArchiveFile.java:86)
	at com.mucommander.sevenzipjbindings.SevenZipJBindingROArchiveFile.check(SevenZipJBindingROArchiveFile.java:76)
	at com.mucommander.commons.file.archive.rar.RarFormatProvider.getFile(RarFormatProvider.java:56)
	at com.mucommander.commons.file.FileFactory.wrapArchive(FileFactory.java:682)
	at com.mucommander.commons.file.FileFactory.getFile(FileFactory.java:491)
	at com.mucommander.commons.file.FileFactory.getFile(FileFactory.java:402)
	at com.mucommander.commons.file.FileFactory.getFile(FileFactory.java:356)
	at com.mucommander.commons.file.FileFactory.getFile(FileFactory.java:314)
	at com.mucommander.sevenzipjbindings.multivolume.SevenZipRarMultiVolumeCallbackHandler.getStream(SevenZipRarMultiVolumeCallbackHandler.java:58)
	at com.mucommander.sevenzipjbindings.SevenZipJBindingROArchiveFile.openInArchive(SevenZipJBindingROArchiveFile.java:86)
	at com.mucommander.sevenzipjbindings.SevenZipJBindingROArchiveFile.check(SevenZipJBindingROArchiveFile.java:76)
	at com.mucommander.commons.file.archive.rar.RarFormatProvider.getFile(RarFormatProvider.java:56)
	at com.mucommander.commons.file.FileFactory.wrapArchive(FileFactory.java:682)
	at com.mucommander.commons.file.FileFactory.getFile(FileFactory.java:491)
	at com.mucommander.commons.file.FileFactory.getFile(FileFactory.java:402)
	at com.mucommander.commons.file.FileFactory.getFile(FileFactory.java:356)
	at com.mucommander.commons.file.FileFactory.getFile(FileFactory.java:314)
	at com.mucommander.sevenzipjbindings.multivolume.SevenZipRarMultiVolumeCallbackHandler.getStream(SevenZipRarMultiVolumeCallbackHandler.java:58)
	at com.mucommander.sevenzipjbindings.SevenZipJBindingROArchiveFile.openInArchive(SevenZipJBindingROArchiveFile.java:86)
	at com.mucommander.sevenzipjbindings.SevenZipJBindingROArchiveFile.check(SevenZipJBindingROArchiveFile.java:76)
	at com.mucommander.commons.file.archive.rar.RarFormatProvider.getFile(RarFormatProvider.java:56)
	at com.mucommander.commons.file.FileFactory.wrapArchive(FileFactory.java:682)
... keeps on going...```

ShayArtzi replied:
That's why I needed to "open up" `FileFactory::createRawFile` which was `private`

ShayArtzi replied:
I tried it and it worked! I restored `FileFactory::createRawFile` access modifier back to `private` as well

ahadas commented on sevenzipjbindings/src/main/java/com/mucommander/sevenzipjbindings/multivolume/InArchiveWrapper.java:
```diff
@@ -0,0 +1,137 @@
+                try {
+                    c.close();
+                } catch (Exception e) {
+                    e.printStackTrace();
```
should use logger instead

ShayArtzi replied:
done

ahadas commented:
@ShayArtzi thanks for posting this!

> Some points I'd like to get your thought on if possible:
> 
> 1. I created `SevenZipRarMultiVolumeCallbackHandler` to handle multi volume archives and used a regular expression on the file name to detect when it's right to use it. This can be optimized by not compiling the pattern each time but keeping it as a field in the class. What do you think about that approach?

it's not that important but with file searches but as it doesn't take much to cache the pattern it would be better, we can open quite a few files now very frequently with file searches

> 2. I wanted to make sure all volume streams are closed at the end of reading operation, so I created a wrapper class `InArchiveWrapper` that carries over the original `IInArchive` and the handler that references other volume streams.

yeah, noticed that - looks good

> 3. I needed a way to create an `AbstractFile` without recursively calling `SevenZipRarMultiVolumeCallbackHandler::getStream`. I couldn't find a proper way to do it so I changed the modifier of `FileFactory::createRawFile` to `public`. I guess that's not ideal so I'm open to other suggestions (I wanted to reuse the existing `SignatureCheckedRandomAccessFile` class)

commented on that inside, maybe I miss something here but I think it should be possible to find a FileFactory#getFile method to call instead

ShayArtzi commented:
@ahadas I think I addressed all issues except for point number 3. See the stack trace I have attached - basically the way `FileFactory::getFile` works takes me into an endless recursion - if you have a better suggestion then my implementation I'm willing to give it a shot - let me know what you think :)

ahadas reviewed (approved):
@ShayArtzi thanks for this contribution!
